pygmt/x2sys.py

```diff
@@ -0,0 +1,142 @@
+
+    fmtfile : str
+        Format definition file prefix for this data set [See Format Definition
+        Files for more information]. Specify full path if the file is not in
+        the current directory.
+
+        Some file formats already have definition files premade. These include:
+
+        - mgd77 (for plain ASCII MGD77 data files)
+        - mgd77+ (for enhanced MGD77+ netCDF files)
+        - gmt (for old mgg supplement binary files)
+        - xy (for plain ASCII x, y tables)
+        - xyz (same, with one z-column)
+        - geo (for plain ASCII longitude, latitude files)
+        - geoz (same, with one z-column).
+
+    """
+    with Session() as lib:
+        arg_str = " ".join([tag, build_arg_string(kwargs)])
+        lib.call_module(module="x2sys_init", args=arg_str)
+
+
+@fmt_docstring
+@use_alias(T="tag", Q="coe")
+def x2sys_cross(tracks=None, outfile=None, **kwargs):
+    """
+    Calculate crossovers between track data files.
+
+    x2sys_cross is used to determine all intersections between ("external
+    cross-overs") or within ("internal cross-overs") tracks (Cartesian or
+    geographic), and report the time, position, distance along track, heading
+    and speed along each track segment, and the crossover error (COE) and mean
+    values for all observables. The names of the tracks are passed on the
+    command line. By default, x2sys_cross will look for both external and
+    internal COEs. As an option, you may choose to project all data using one
+    of the map-projections prior to calculating the COE.
+
+    Full option list at :gmt-docs:`supplements/x2sys/x2sys_cross.html`
+
+    {aliases}
+
+    Parameters
+    ----------
+    tracks : pandas.DataFrame or str
+        Either a table with (x, y) or (lon, lat) values in the first two
+        columns, or a filename (e.g. csv, txt format). More columns may be
+        present.
+
+    outfile : str
+        Optional. The file name for the output ASCII txt file to store the
+        table in.
+
+    tag : str
+        Specify the x2sys TAG which identifies the attributes of this data
+        type.
+
+    coe : str
+        Use **e** for external COEs only, and **i** for internal COEs only
+        [Default is all COEs].
+
+    Returns
+    -------
+    crossover_errors : pandas.DataFrame or None
+        Table containing crossover error information.
+        Return type depends on whether the outfile parameter is set:
+
+        - pandas.DataFrame table with (x, y, ..., etc) if outfile is not set
+        - None if outfile is set (track output will be stored in outfile)
+
+    """
+    # kinds = [data_kind(data=track) for track in tracks]
+    track = tracks[0]
+    kind = data_kind(track)
+
+    with Session() as lib:
+        # for kind, track in zip(kinds, tracks):
+        if kind == "file":
+            file_context = dummy_context(track)
+        elif kind == "matrix":
+            file_context = lib.virtualfile_from_matrix(track.values)
```

Comment:
Making a virtualfile from a `pandas.DataFrame` and passing that into `x2sys_cross` doesn't work yet. It might just be because I'm using `ASCII` with `x2sys_init`, not sure if `BINARY` would work (see https://docs.generic-mapping-tools.org/6.1/supplements/x2sys/x2sys_init.html#format-definition-files). Error message is as follows:

```python-traceback
x2sys_cross [WARNING]: No time column, use dummy times
Not a valid unit: c [meter assumed]
Not a valid unit: c [meter assumed]
x2sys_cross [ERROR]: x2sys_read_file : Cannot find track @GMTAPI@-S-I-D-M-T-N-000000
(null): Error from fclose [@GMTAPI@-S-I-D-M-T-N-000000]
```

Reply:
As I understand how GMT works, I think x2sys related modules don't support virtual files. Perhaps you could open an issue in the GMT repository.

Reply:
Not surprised actually, I'll open an issue!